### DISCUSSION TITLE
nextjs 15 SSR false not working as expected

### DISCUSSION THREAD

**Discussion by sebalaini** [vercel/next.js#72236] [Help]
### Summary

I'm upgrading my Next.js 14 app to V15 but have problems with the `ssr: false` option.

in my `layout.tsx` file I have this:
```js
...
const CookieBanner = dynamic(() => import('@repo/ui/CookieBanner'), {
  ssr: false,
})
...

export default function RootLayout({ children }: { children: ReactNode }) {
  return (
    <html
      lang='en'
      className='scroll-smooth'
    >
       ...
        <CookieBanner
          variant='bg-purple-500'
       ...
```

and the component is like so:
```js
'use client'
...
const CookieBanner = ({ variant, cookieImg, text, cta }: CookieBannerProps) => {
...
  window.addEventListener(
    'scroll',
    () => {
      checkCookie()
    },
    { once: true }
  )

  return (
...

```

this works perfectly in V14 but when I upgrade to V15 it doesn't compile, now I reverted to V14 but it's something about not being able to use `ssr: false` in server components, I know I can remove that option but by reading the doc it should be possible to use it still.

One of the reasons I want to use `ssr: false` beside performance is that I have warnings about `window` being `undefined`.
This is not just for the cookie banner but for the Google map as well, which by using this option allows me to have a better lighthouse score.

**Comment by ioanvranau** [comment]
I had also this problem. I moved all to ssr: true .. but I think I made a mistake.

**Comment by jjmean2** [comment]
I encountered a similar issue while using a Lit component wrapped with [createComponent from @lit/react](https://lit.dev/docs/frameworks/react/#createcomponent). Here’s an example setup:

```tsx
'use client'

import React from 'react';
import { createComponent } from '@lit/react';
import { MyElement } from './my-element.js';

const MyElementComponent = createComponent({
  tagName: 'my-element',
  elementClass: MyElement,
  react: React,
  events: {
    onactivate: 'activate',
    onchange: 'change',
  },
});

export default MyElementComponent;
```

I dynamically import this component in a Server Component like this:
```tsx
const MyElementComponent = dynamic(() => import('./MyElementComponent'), { ssr: false });

async function ServerComponent() {
  ...
  return (
    <div>
      <MyElementComponent />
      ...
    </div>
  );
}
```

Since `createComponent` relies on `HTMLElement` from the `window` object—which is only available in the browser—marking `ssr: false` is necessary to prevent errors during SSR.

This worked perfectly in Next.js 14.

However, after updating to Next.js 15, using `ssr: false` now causes a build error.

To workaround this issue, I had to create a thin wrapper client component solely for marking `ssr: false`:

```tsx
'use client';

const MyElementComponent = dynamic(() => import('./MyElementComponent'), { ssr: false });

// Thin wrapper client component just for ssr: false
export function MyElementComponentWrapper(props) {
  return <MyElementComponent {...props} />;
}
```
Then, I use the wrapper in my Server Component:
```tsx
import { MyElementComponentWrapper } from './MyElementComponentWrapper';

async function ServerComponent() {
  ...
  return (
    <div>
      <MyElementComponentWrapper />
      ...
    </div>
  );
}
```

I find this approach cumbersome, as it adds unnecessary boilerplate to handle something that worked seamlessly in previous versions.

Thank you for your time and consideration. I’d greatly appreciate it if this could be addressed.

**Reply by sebalaini** [reply]
Yeah, I was thinking of doing something similar, having a useless extra component only to call that `ssr: false` component, but it looked like a bad idea. Why should I add extra stuff when it was working perfectly in the previous version?

This looks like a problem they have to fix, not something we need to create a workaround because they changed the behavior.

**Reply by Konadu-Akwasi-Akuoko** [reply]
Thanks @jjmean2 , been battling this issue for like 4hrs straight, and all I had to do was wrap the dynamic import in a client component.

Seriously I am weird, but Nextjs is becoming weider than me

**Reply by servertdot** [reply]
also like a variant use `!!false` :)

```tsx
const HistoryModal = dynamic(() => import('./parts/history-modal'), {
    ssr: false,
});
```
<img width="637" alt="Screenshot 2025-02-06 at 11 04 35" src="https://github.com/user-attachments/assets/dded614b-b6de-474b-853d-a24a64898d01" />


if i made `!!false` - i have no error and not need a new wrapper for component

```tsx
const HistoryModal = dynamic(() => import('./parts/history-modal'), {
    ssr: !!false,
});
```

but anyway it's shit

**Comment by DiegoBistro** [comment]
same issue here.

Has anyone found a solution?

**Comment by roelkok** [comment]
I'm using PaperJS which throws an error when used in a server build.
Setting `ssr: false` worked for dev, but for the build the error came back.

I eventually was able to work around it by setting the Next config to resolving the PaperJS import to nothing on the server.
As described here https://github.com/paperjs/paper.js/issues/2050#issuecomment-2059150214

```
const nextConfig = {
  // ...other settings
  webpack: (config, { isServer }) => {
    if (isServer) {
      config.resolve.alias['paper'] = false;
    }
    return config;
  },
};
```
